Initial commit

ionkoto opened the pull request.

ionkoto committed:
Initial commit

ionkoto committed:
Root folder rename.
Added some documentation.

alex-kostadinov commented on com.musala.simple.students.db/pom.xml:
```diff
@@ -0,0 +1,71 @@
+							<transformers>
+								<transformer
+									implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+									<mainClass>com.musala.simple.students.Main</mainClass>
```
Isn't this the main class from the other project? :)

ionkoto replied:
Haha good point 👍

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java:
```diff
@@ -0,0 +1,189 @@
+	public Database getDatabase() {
+		Database database = null;
+
+		switch (this.getDbType()) {
```
Formatting ;)

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java:
```diff
@@ -0,0 +1,189 @@
+	 * 
+	 * @return a new database of the required type
+	 */
+	public Database getDatabase() {
```
The factory is fine, but we can make it better! :) Let's change this method to be named `public static Database createDatabase(DatabaseTypes dbType)` and implement the so-called Builder pattern (often supported by factories). Perform the following steps:

- Remove the implemented constructor, we won't need it.
- The static method `createDatabase(DatabaseTypes dbType)` (previously `getDatabase()`) should internally do the currently implemented logic with the switch statement and return the newly created `Database` object (MySQL database, Mondo database, etc).
- Then, all setters should be renamed and re-declared too - e.g. `setName()` has to become `withName(String name)` and internally the logic should be - we call `withName(dbProperties.getProperty(DB_NAME))` and underneath it does `this.name = name` and then `return this`.
- Keep in mind that making `createDatabase()` static will require all with-something methods to be static.
- Finally, you will have to have a `public Database build()` method which returns `this` instance.
- The getter methods, on the other hand, can be removed - they don't seem to be used anywhere.

The whole point is that, externally, I can do the following call:

```java
Properties dbProperties = FileHelper.readDbPropertiesFile(DatabaseTypes.MongoDb);
Database database = DatabaseFactory.createDatabase(DatabaseTypes.MongoDb).withName(dbProperties.getProperty(DB_NAME))withHost(dbProperties.getProperty(DB_HOST)).withPort(...).withUsername(...).withPassword().build().
```

You can use this github class as a reference: https://github.com/eclipse/smarthome/blob/4204ce06bb28c28e5f711e720f87ef83beff2e27/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/builder/ThingBuilder.java

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java:
```diff
@@ -0,0 +1,258 @@
+	private MongoDatabase database;
+	private MongoCollection<Document> studentsCollection;
+	private String dbName;
+	private Logger logger = LoggerFactory.getLogger(Main.class);
```
Every time you declare a logger, it has to be bound to the current class (this.getClass()). In such way, logs will be printed as being emitted by this class. If you leave this logger with `getLogger(Main.class)`, logs from `MyMongoDatabase` will be listed as logged by the `Main` which is incorrect. Please correct this everywhere.

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java:
```diff
@@ -0,0 +1,258 @@
+
+	@Override
+	public void deleteStudentById(int studentId) {
+
```
Why there is no implementation for this method? :)

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java:
```diff
@@ -0,0 +1,258 @@
+
+		List list = new ArrayList(query.values());
+		int id = (int) list.get(1);
+		String name = (String) list.get(2);
```
Everywhere you cast directly (String), replace with `Objects.toString(list.get(2))`. In such case you make use of Java 8's built-in anti-null-pointer String prevention. :) Read more about `Objects.toString()`.

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/MyMongoDatabase.java:
```diff
@@ -0,0 +1,258 @@
+
+				Document doc = cur.next();
+
+				List list = new ArrayList(doc.values());
```
The logic implemented between lines 200 and 204 is used in both `getStudentById()` and `getAllStudentsArr()` methods - extract it in an external method and call it in both places.

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/helpers/FileHelper.java:
```diff
@@ -0,0 +1,87 @@
+ * 
+ */
+public class FileHelper {
+	private static Logger logger = LoggerFactory.getLogger(Main.class);
```
Correct logger.

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Main.java:
```diff
@@ -0,0 +1,78 @@
+
+public class Main {
+	public static void main(String[] args) {
+		Logger logger = LoggerFactory.getLogger(Main.class);
```
Here the logger declaration should remain as it is! ;)

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Main.java:
```diff
@@ -0,0 +1,78 @@
+public class Main {
+	public static void main(String[] args) {
+		Logger logger = LoggerFactory.getLogger(Main.class);
+		BasicConfigurator.configure();
```
What does log4j has to do with slf4j? Do we need this statement and `import org.apache.log4j.BasicConfigurator;` at all?

alex-kostadinov reviewed (changes requested):
That's it for now, please correct the changes. Folder migration has worked fine! :)

alex-kostadinov commented on com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/Database.java:
```diff
@@ -0,0 +1,77 @@
+package com.musala.simple.students.db;
```
Do the followint package changes:
 - `com.musala.simple.students.db.database` - in this new package the `Database`, `DatabaseFactory` and `DatabaseTypes` classes should be moved.
 - `com.musala.simple.students.db.database.impl` - in this new package the Mongo and other implementation classes should be moved.

ionkoto committed:
DatabaseFactory refactoring
Database abstraction refactoring

ionkoto committed:
Package naming fix

DatabaseProperties - javadoc added

Database interface renamed to AbstractDatabase, javadoc added

DatabaseFactory > createDatabase > default switch case > logg info added

Other final fixes.

Readme added for last two projects.
